### DISCUSSION TITLE
Solo & mute features

### DISCUSSION THREAD

**Discussion by MrDrone-ru** [CoderLine/alphaTab#1024] [Q&A] [answered]
Hi everyone!
Is it true that now the variables `isSolo` or `isMute` do not affect the playback of tabs, and the logic of their work should be controlled through a variable `volume`? Are there any plans to add this logic to AT in the future?

**Comment by Danielku15** [comment, marked as answer]
Hi. 
Yes, `isSolo` and `isMute` are not affecting directly by default the playback and this is rather by design to avoid unexpected side effects for people who do not build a full UI around the mute/solo functionalities.  

You can simply use the provided APIs for muting and solo tracks to update the player. to respect the information in the tracks. 

https://alphatab.net/docs/reference/api/changetrackmute/
https://alphatab.net/docs/reference/api/changetracksolo/

```js
api.changeTrackSolo(score.tracks.filter(t => t.playbackInfo.isSolo), true);
api.changeTrackMute(score.tracks.filter(t => t.playbackInfo.isMute), true);
```

**Reply by MrDrone-ru** [reply]
Ok, i got it. 
But there is one more noob question - about volume control.
According to the docs there are 2 functions to change the volume.
1. API.changeTrackVolume(tracks[], x) where x is multiplier 0<x<3
Let's say I have a track with volume=3 and i want I need to set the maximum volume for it. How should I do it?
From my tests, it turns out that initially such a track is not audible at all. 
This
`api.changeTrackVolume([api.score.tracks[0]], 1);`
 gives base volume = 3.
 and that `api.changeTrackVolume([api.score.tracks[0]], 3);` fits volume=9 right?
but I doubt that this is what I need, given that changing the parameter `track.volume` does not affect the operation of this function.
2. AlphaSynth.SetChannelVolume(int channel, double volume)
This is probably what I need, but I did not find how to call it.

**Reply by Danielku15** [reply]
`changeTrackVolume `allows you to control the volume of a track using the main song model while `setChannelVolume `controls the internal midi volume. `changeTrackVolume `calls internally `setChannelVolume` with the midi channels the track uses so there is no difference.

Volume control works like this: Each track has a defined base volume. This is usually defined by the author of the file. This volume is stored in `score.tracks[].playbackInformation` and will end up in the generated Midi.  

The Track/Channel volume feature is a multiplier on top of this volume to allow "boosting" (or damping) which is considered when mixing the audio together. Internally the audio is synthesized using a midi+soundfont synthesizer which is organized in channels. On each channel notes can be played and each channel is configured with details like instrument and base volume. The audio samples for each channel are generated and then mixed together to get an overall audio wave for playback. During this sample generation and mixing the volumes are acting as multiplier.

If you have one track with volume 3 and another with 16 and you want to make them sound the same (volume 16) you would set the track volume of the first track to 5.3 to bring them to the same volume level.

**Reply by MrDrone-ru** [reply]
So there is an error in the help, and x can take values up to 16. 
But what if the initial volume of the track is zero? It will be impossible to increase its volume, right?
I would like to directly set the volume of a track or channel from 0 to 16, is this not implemented at the moment?

**Reply by Danielku15** [reply]
Yes, if it is set to zero then logically multiplying will not have any effect. In such cases you would change the volume of the tracks directly before giving it to alphaTab for further usage. Dynamic change of the track volume in the data model has no effect. The midi file is generated only once on initial load. 

You would likely load the files manually in such cases and let it parse it into a `Score`: https://alphatab.net/docs/guides/lowlevel-apis#loading-files-via-scoreloader Then you simply loop through the tracks and modify the `playbackInfo.volume` to your desired arrangement, And then you pass it into the API through `renderScore` or `renderTracks`. 

Regarding the docs they really need to be updated. The limitation was removed as part of https://github.com/CoderLine/alphaTab/pull/487 in the 1.2 release.
